File: src/main/java/com/github/bjoernpetersen/jmusicbot/playback/WeakQueueChangeListener.java
package com.github.bjoernpetersen.jmusicbot.playback;

import java.lang.ref.WeakReference;
import java.util.Optional;
import javax.annotation.Nonnull;

public final class WeakQueueChangeListener implements QueueChangeListener {

  @Nonnull
  private final WeakReference<QueueChangeListener> wrapped;

  public WeakQueueChangeListener(@Nonnull QueueChangeListener wrapped) {
    this.wrapped = new WeakReference<>(wrapped);
  }

  @Nonnull
  private Optional<QueueChangeListener> getWrapped() {
    return Optional.ofNullable(wrapped.get());
  }

  @Override
  public void onAdd(@Nonnull QueueEntry entry) {
    getWrapped().ifPresent(wrapped -> wrapped.onAdd(entry));
  }

  @Override
  public void onRemove(@Nonnull QueueEntry entry) {
    getWrapped().ifPresent(wrapped -> wrapped.onRemove(entry));
  }
}

File: src/main/java/com/github/bjoernpetersen/jmusicbot/playback/WeakQueueChangeListener.kt
package com.github.bjoernpetersen.jmusicbot.playback

import java.lang.ref.WeakReference
import java.util.*

class WeakQueueChangeListener(wrapped: QueueChangeListener) : QueueChangeListener {

  private val wrapped: WeakReference<QueueChangeListener> = WeakReference(wrapped)

  private fun getWrapped(): Optional<QueueChangeListener> {
    return Optional.ofNullable(wrapped.get())
  }

  override fun onAdd(entry: QueueEntry) {
    getWrapped().ifPresent { wrapped -> wrapped.onAdd(entry) }
  }

  override fun onRemove(entry: QueueEntry) {
    getWrapped().ifPresent { wrapped -> wrapped.onRemove(entry) }
  }

  override fun onMove(entry: QueueEntry, fromIndex: Int, toIndex: Int) {
    getWrapped().ifPresent { it.onMove(entry, fromIndex, toIndex) }
  }
}
